### PR TITLE
Set maximum CPU and Memory requests on K8s

### PR DESCRIPTION
**Pull request by jopadi**
Configure max requests for CPU and Memory.

### CODE REVIEW

File: pkg/cluster/k8sres.go [hunk 1]
```diff
@@ -183,6 +188,51 @@ func (c *Cluster) enforceMinResourceLimits(resources *v1.ResourceRequirements) e
 	return nil
 }
 
+func (c *Cluster) enforceMaxnRequests(resources *v1.ResourceRequirements) error {
```
Comment (FxKu) [inline pkg/cluster/k8sres.go:191]:
```suggestion
func (c *Cluster) enforceMaxRequests(resources *v1.ResourceRequirements) error {
```
n?

File: pkg/cluster/k8sres.go [hunk 2]
```diff
@@ -131,10 +131,15 @@ func makeDefaultResources(config *config.Config) acidv1.Resources {
...
+		CPU:    config.Resources.MaxCPURequest,
+		Memory: config.Resources.MaxMemoryRequest,
+	}
 
 	return acidv1.Resources{
-		ResourceRequests: defaultRequests,
-		ResourceLimits:   defaultLimits,
+		ResourceRequests:    defaultRequests,
+		ResourceLimits:      defaultLimits,
+		MaxResourceRequests: maxRequests,
```
Comment (FxKu) [inline pkg/cluster/k8sres.go:142]:
```suggestion
```
Do not add it here. It's not a manifest option.

File: pkg/cluster/k8sres.go [hunk 3]
```diff
@@ -183,6 +183,51 @@ func (c *Cluster) enforceMinResourceLimits(resources *v1.ResourceRequirements) e
...
 }
 
+func (c *Cluster) enforceMaxRequests(resources *v1.ResourceRequirements) error {
+	var (
+		isMax bool
+		err   error
+		msg   string
+	)
+
+	// setting limits too low can cause unnecessary evictions / OOM kills
```
Comment (FxKu) [inline pkg/cluster/k8sres.go:193]: take care about comments and log messages to talk about the right thing

File: pkg/cluster/k8sres.go [hunk 4]
```diff
@@ -183,6 +183,51 @@ func (c *Cluster) enforceMinResourceLimits(resources *v1.ResourceRequirements) e
 	return nil
 }
 
+func (c *Cluster) enforceMaxRequests(resources *v1.ResourceRequirements) error {
```
Comment (FxKu) [inline pkg/cluster/k8sres.go:186]:
```suggestion
func (c *Cluster) enforceMaxResourceRequests(resources *v1.ResourceRequirements) error {
```

File: pkg/cluster/k8sres.go [hunk 5]
```diff
@@ -183,6 +183,51 @@ func (c *Cluster) enforceMinResourceLimits(resources *v1.ResourceRequirements) e
...
+		if err != nil {
+			return fmt.Errorf("could not compare defined CPU limit %s for %q container with configured maximum value %s: %v",
+				cpuRequest.String(), constants.PostgresContainerName, maxCPURequest, err)
+		}
+		if isMax {
+			msg = fmt.Sprintf("defined CPU limit %s for %q container is below required maximum %s and will be increased",
+				cpuRequest.String(), constants.PostgresContainerName, maxCPURequest)
+			c.logger.Warningf(msg)
+			c.eventRecorder.Eventf(c.GetReference(), v1.EventTypeWarning, "ResourceLimits", msg)
+			resources.Limits[v1.ResourceCPU], _ = resource.ParseQuantity(maxCPURequest)
```
Comment (FxKu) [inline pkg/cluster/k8sres.go:207]:
```suggestion
			resources.Requests[v1.ResourceCPU], _ = resource.ParseQuantity(maxCPURequest)
```

File: pkg/cluster/k8sres.go [hunk 6]
```diff
@@ -183,6 +183,51 @@ func (c *Cluster) enforceMinResourceLimits(resources *v1.ResourceRequirements) e
...
+		if err != nil {
+			return fmt.Errorf("could not compare defined memory limit %s for %q container with configured maximum value %s: %v",
+				memoryRequest.String(), constants.PostgresContainerName, maxMemoryRequest, err)
+		}
+		if isMax {
+			msg = fmt.Sprintf("defined memory limit %s for %q container is below required maximum %s and will be increased",
+				memoryRequest.String(), constants.PostgresContainerName, maxMemoryRequest)
+			c.logger.Warningf(msg)
+			c.eventRecorder.Eventf(c.GetReference(), v1.EventTypeWarning, "ResourceLimits", msg)
+			resources.Limits[v1.ResourceMemory], _ = resource.ParseQuantity(maxMemoryRequest)
```
Comment (FxKu) [inline pkg/cluster/k8sres.go:224]:
```suggestion
			resources.Requests[v1.ResourceMemory], _ = resource.ParseQuantity(maxMemoryRequest)
```

File: pkg/util/util.go [hunk 7]
```diff
@@ -367,3 +367,19 @@ func IsSmallerQuantity(requestStr, limitStr string) (bool, error) {
...
+		return false, fmt.Errorf("could not parse request %v : %v", requestStr, err)
+	}
+
+	limit, err2 := resource.ParseQuantity(limitStr)
+	if err2 != nil {
+		return false, fmt.Errorf("could not parse limit %v : %v", limitStr, err2)
+	}
+
+	return request.Cmp(limit) == 1, nil
+}
```
Comment (FxKu) [inline pkg/util/util.go:385]: as mentioned above, we can cover everything with just one comparison function which is also covered by a unit test

**Comment by Jan-M** [comment]
how do we deal with the conflict now of maxmemory vs minmemorylimit? can also be startup error if they conflict

**Comment by FxKu** [comment]
:+1:

**Comment by jopadi** [comment]
👍

**Pull request merged**